### PR TITLE
Add modkit pileup

### LINKED ISSUE #3950: new module: modkit/pileup
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by fellen31**
## PR checklist

Closes #3950,   continuation of #4490 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/modkit/pileup/main.nf [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    tuple val(meta), path("*.bed")     , emit: bed     , optional: true
+    tuple val(meta), path("*.bedgraph"), emit: bedgraph, optional: true
+    tuple val(meta), path("*.log")     , emit: log     , optional: true
+    path "versions.yml"                , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
```
Comment (itrujnara) [inline modules/nf-core/modkit/pileup/main.nf:25]: Align the equal signs

File: modules/nf-core/modkit/pileup/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+    test("[bam, bai], [], []") {
+
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id: 'bc_anchored' ],
+                    file('/media/ssd_4tb/felix/projects/fellen31/test-datasets/data/delete_me/modkit/bc_anchored_10_reads.haplotyped.sorted.bam', checkIfExists: true),
+                    file('/media/ssd_4tb/felix/projects/fellen31/test-datasets/data/delete_me/modkit/bc_anchored_10_reads.haplotyped.sorted.bam.bai', checkIfExists: true)
```
Comment (itrujnara) [inline modules/nf-core/modkit/pileup/tests/main.nf.test:19]: The tests are very comprehensive, but you cannot just link data from your personal portable drive, as nobody (especially not AWS runners) will run the tests with your hardware. Consult #5281 for an example of how to use test data in module tests.

File: tests/modules/nf-core/modkit/pileup/main.nf [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env nextflow
```
Comment (itrujnara) [inline tests/modules/nf-core/modkit/pileup/main.nf:1]: This seems to be related to Pytest. We exclusively use nf-test now, so there is no need for these files.

**Review by itrujnara — COMMENTED** [review]
Thanks for adding this module. Overall good job, but there are some important issues to address. You need to check test data paths, and you DO NOT need to add pytest, as we only use nf-test now.

File: modules/nf-core/modkit/pileup/tests/main.nf.test (hunk 2 above)
Comment (fellen31) [reply modules/nf-core/modkit/pileup/tests/main.nf.test:19]: Hi, thanks for the review. I'm waiting for https://github.com/nf-core/test-datasets/pull/1224, that's why I kept the PR as a draft. 

Will update after the testdata PR is reviewed and merged.

File: tests/modules/nf-core/modkit/pileup/main.nf (hunk 3 above)
Comment (fellen31) [reply tests/modules/nf-core/modkit/pileup/main.nf:1]: Good spot!

**Review by SPPearce — APPROVED** [review]
Looks good, lots of tests there.
Does at least one of the tests check that the folder moving part works? It looks like the tool sometimes produces a folder and sometimes a file, are you testing both bits?

**Comment by fellen31** [comment]
> Looks good, lots of tests there. Does at least one of the tests check that the folder moving part works? It looks like the tool sometimes produces a folder and sometimes a file, are you testing both bits?

I saw now I had actually missed adding --partition-tag HP to two of the test configs, and --prefix to the module...
But yes, after this update:

### no config or --preset-traditional
- Produces one bed file `prefix.tmp` which becomes `prefix.bed` if prefix.tmp is not a directory

### --preset traditional --partition-tag HP
- A directory with multiple bed files, one per haplolag + ungrouped
- e.g. `prefix.tmp/prefix_1.bed` which becomes `prefix_1.bed` 

###  --preset traditional --partition-tag HP --bedgraph
- A directory with multiple bedgraph files, one per haplolag-modification combination
- e.g. `prefix.tmp/prefix_1_m_CG0_combined.bedgraph` which becomes `prefix_1_m_CG0_combined.bedgraph`

**Review by SPPearce — APPROVED** [review]

**Pull request merged**